app/flatpak-builtins-uninstall.c

```diff
@@ -127,8 +127,15 @@ find_used_refs (FlatpakDir *dir, GHashTable *used_refs, const char *ref, const c
     {
       FlatpakRelated *rel = g_ptr_array_index (related, i);
 
-      if (!rel->auto_prune)
-        g_hash_table_add (used_refs, g_strdup (rel->ref));
+      if (!rel->auto_prune && !g_hash_table_contains (used_refs, rel->ref))
+        {
+          g_autofree char *related_origin = NULL;
+          related_origin = flatpak_dir_get_origin (dir, rel->ref, NULL, NULL);
+          if (related_origin != NULL)
+            find_used_refs (dir, used_refs, rel->ref, related_origin);
+
+          g_hash_table_add (used_refs, g_strdup (rel->ref));
```

Comment:
I think you need to add to used_ref before recursing, because otherwise you could accidentally get stuck in a loop in case some related ref is related to the original ref.

Reply:
Good point, pushed a fixup